subprojects/messaging/src/main/java/org/gradle/internal/event/DefaultListenerManager.java

```diff
@@ -64,10 +64,14 @@ public void removeListener(Object listener) {
             ListenerDetails details = allListeners.remove(listener);
             if (details != null) {
                 details.disconnect();
-                for (EventBroadcast<?> broadcaster : broadcasters.values()) {
-                    broadcaster.maybeRemove(details);
+                details.listenerLock.lock();
```

Comment:
❌  I think there is a remote race condition here.  Something like:

Thread 1: call removeListener
Thread 2: sends broadcast event
Thread 1: acquires `DefaultListenerManager.lock`
Thread 2: acquires `details.listenerLock` before Thread 1 can
Thread 2: listener then tries to do something with DefaultListenerManager such as adding a listener for a different notification or calling `getBroadcaster()`

The timing would have to be just right, but I found at least [one case](https://github.com/gradle/gradle/blob/0946ab90bec3b5bd6e135504faf2cebcc1feaedd/subprojects/launcher/src/main/java/org/gradle/launcher/daemon/server/scaninfo/DefaultDaemonScanInfo.java#L83-L83) where we were calling back into the listener manager from a listener, so I don't think this is that far fetched. 

Perhaps we should do this in two steps.  First, remove from allListeners and call details.disconnect().  This ensures that the listener will no longer be executed.  Then we release the listener manager lock and do the `maybeRemove` separately perhaps only getting the broadcaster dispatch lock.  For that matter, I wonder if the `EventBroadcaster` methods that require a lock should be locking around the dispatch lock now instead of the listener manager lock.